src/api/object.js

```diff
@@ -1,15 +1,71 @@
 'use strict'
 
 const argCommand = require('../cmd-helpers').argCommand
+const DAGNode = require('ipfs-merkle-dag').DAGNode
 
 module.exports = (send) => {
   return {
     get: argCommand(send, 'object/get'),
-    put (file, encoding, cb) {
-      if (typeof encoding === 'function') {
-        return cb(null, new Error("Must specify an object encoding ('json' or 'protobuf')"))
+    put (obj, options, callback) {
+      if (typeof options === 'function') {
+        callback = options
+        options = {}
       }
-      return send('object/put', encoding, null, file, cb)
+      if (!options) {
+        options = {}
+      }
+
+      let tmpObj = {
+        Data: null,
+        Links: []
+      }
+
+      if (Buffer.isBuffer(obj)) {
+        if (!options.enc) {
+          tmpObj = { Data: obj.toString(), Links: [] }
+        } else {
+          tmpObj = JSON.parse(obj.toString())
+        }
+      } else if (obj.multihash) {
+        tmpObj = {
+          Data: obj.data.toString(),
+          Links: obj.links.map((l) => { return l.toJSON() })
+        }
+      } else if (typeof obj === 'object') {
+        tmpObj.Data = obj.Data.toString()
+      } else {
+        return callback(new Error('obj not recognized'))
+      }
+
+      const buf = new Buffer(JSON.stringify(tmpObj))
+      const enc = options.enc || 'json'
+
+      return send('object/put', enc, null, buf, (err, result) => {
+        if (err) {
+          return callback(err)
+        }
+
+        if (Buffer.isBuffer(obj)) {
+          if (!options.enc) {
+            obj = { Data: obj, Links: [] }
+          } else {
+            obj = JSON.parse(obj.toString())
+          }
+        }
+        let node
+        if (obj.multihash) {
+          node = obj
+          console.log('->', node)
+        } else {
+          node = new DAGNode(obj.Data, obj.Links)
+        }
+
+        if (node.toJSON().Hash !== result.Hash) {
+          return callback(new Error('Stored object was different from constructed object'))
+        }
+
+        callback(null, node)
```

Comment:
In order to support the variety of ways `ipfs.object.put` can be called, I had to break the elegant promise injection that was there before. @dignifiedquire wanna share some wisdom on how we can add promises again, in the most elegant way possible?